### PR TITLE
Feature airfoil transition

### PR DESCRIPTION
**Pull request by rachealerhard**
This adds a script for generating transitional airfoil geometries interpolated between two given airfoils

### CODE REVIEW

File: regression/scripts/airfoil_import/generate_airfoil_transitions.py [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    a2 = airfoils_path + a_labels[1]+ ".txt"
+    new_files = generate_airfoil_transition(a1, a2, space, nairfoils, save_file=True, save_filename="Polars/Transition")
+    #plt.show()
+    
+    # open and check new file:
+    f = open(new_files[next(iter(new_files))].name)
+    data_block = f.readlines()
+    
+    test_entry = data_block[30]
+    assert( test_entry == ' 0.514476 0.080181\n')
```
Comment (planes) [inline regression/scripts/airfoil_import/generate_airfoil_transitions.py:30]: I would do a full check of all points. Have a regression saved file, import both files as arrays, subtract one from the other and make sure the max error is below a tolerance.

File: trunk/SUAVE/Methods/Geometry/Two_Dimensional/Cross_Section/Airfoil/generate_airfoil_transitions.py [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+#  Imports
+# ----------------------------------------------------------------------
+
+from SUAVE.Methods.Geometry.Two_Dimensional.Cross_Section.Airfoil.import_airfoil_geometry import import_airfoil_geometry 
+from SUAVE.Plots.Geometry_Plots import plot_airfoil
+import numpy as np
+
+def generate_airfoil_transition(a1, a2, space, nairfoils, save_file=False, save_filename="Transition", npts=50):
+    """ Takes in two airfoils and their locations and interpolates
+    between them to establish a transition region.
```
Comment (planes) [inline trunk/SUAVE/Methods/Geometry/Two_Dimensional/Cross_Section/Airfoil/generate_airfoil_transitions.py:17]: can you find a better word than transition? When I see airfoil transition I just think laminar/turbulent.

maybe just airfoil_interpolation

File: trunk/SUAVE/Methods/Geometry/Two_Dimensional/Cross_Section/Airfoil/generate_airfoil_transitions.py [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+            
+            
+    # save new airfoil geometry files:
+    if save_file:
+        new_files = {'a_{}'.format(i+1): [] for i in range(nairfoils-2)}
+        airfoil_names = []
+
+        for k in range(nairfoils-2):
+            # create new files and write title block for each
+            title_block     = "Airfoil Transition "+str(k+1)+" between"+a1+"and"+a2+"\n 61. 61.\n\n"
```
Comment (planes) [inline trunk/SUAVE/Methods/Geometry/Two_Dimensional/Cross_Section/Airfoil/generate_airfoil_transitions.py:79]: does 61 mean the number of points?

File: trunk/SUAVE/Methods/Geometry/Two_Dimensional/Cross_Section/Airfoil/generate_airfoil_transitions.py [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+    between them to establish a transition region.
+    
+    Assumptions: Linear geometric transition between airfoils
+    
+    Source: None
+    
+    Inputs:
+    a1                 first airfoil                                [ airfoil ]
+    a2                 second airfoil                               [ airfoil ]
+    y1                 spanwise location of first airfoil           [ m ]
```
Comment (planes) [inline trunk/SUAVE/Methods/Geometry/Two_Dimensional/Cross_Section/Airfoil/generate_airfoil_transitions.py:26]: y1 and y2 are  not used

File: trunk/SUAVE/Methods/Geometry/Two_Dimensional/Cross_Section/Airfoil/generate_airfoil_transitions.py [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+    a_geo = import_airfoil_geometry(airfoil_geo_files,npts)
+    
+    # identify x and y coordinates of the two airfoils
+    x_upper = a_geo.x_upper_surface
+    y_upper = a_geo.y_upper_surface
+    x_lower = a_geo.x_lower_surface
+    y_lower = a_geo.y_lower_surface
+    
+    # setup spanwise stations for transitional airfoils
+    z = np.linspace(0,space,nairfoils)
```
Comment (planes) [inline trunk/SUAVE/Methods/Geometry/Two_Dimensional/Cross_Section/Airfoil/generate_airfoil_transitions.py:44]: I don't understand this space variable.

For example, in the regression it has airfoils from 0 to 0.2 with 4 points so: 0, 0.067, .133, and 0.2

But then  zairfoils is 0 to space. So what does space do? I set space to 1.0 and it gave me the same result.

File: trunk/SUAVE/Methods/Geometry/Two_Dimensional/Cross_Section/Airfoil/generate_airfoil_transitions.py [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+    xairfoils_upper  = np.array(x_upper).T
+    xairfoils_lower  = np.array(x_lower).T
+
+    
+    # generate new upper and lower airfoil data for transitional airfoils
+    upper_data = {'a_{}'.format(i+1): [] for i in range(nairfoils-2)}
+    lower_data = {'a_{}'.format(i+1): [] for i in range(nairfoils-2)}
+    
+    # for each point around the airfoil, interpolate between the two given airfoils
+    for i in range(npts):
```
Comment (planes) [inline trunk/SUAVE/Methods/Geometry/Two_Dimensional/Cross_Section/Airfoil/generate_airfoil_transitions.py:59]: This is something that should be vectorizable.

File: trunk/SUAVE/Plots/Geometry_Plots/plot_airfoil.py [hunk 7]
```diff
@@ -33,20 +33,38 @@ def plot_airfoil(airfoil_names,  line_color = 'k-', save_figure = False, save_fi
...
     airfoil_data = import_airfoil_geometry(airfoil_names)       
-
-    for i in range(len(airfoil_names)):
-        # separate x and y coordinates 
-        airfoil_x  = airfoil_data.x_coordinates[i] 
-        airfoil_y  = airfoil_data.y_coordinates[i]    
-
-        name = save_filename + '_' + str(i)
+    
+    if overlay:
```
Comment (planes) [inline trunk/SUAVE/Plots/Geometry_Plots/plot_airfoil.py:37]: I'm not sure if was from this script or the one before. But it spit out a bunch of blank plots.

File: trunk/SUAVE/Plots/Geometry_Plots/plot_airfoil.py [hunk 8]
```diff
@@ -33,20 +33,38 @@ def plot_airfoil(airfoil_names,  line_color = 'k-', save_figure = False, save_fi
...
+            # separate x and y coordinates 
+            airfoil_x  = airfoil_data.x_coordinates[i] 
+            airfoil_y  = airfoil_data.y_coordinates[i]    
+    
+            name = save_filename + '_' + str(i)
+            fig  = plt.figure(name)
+            axes = fig.add_subplot(1,1,1)
+            axes.set_title(airfoil_names[i])
+            axes.plot(airfoil_x, airfoil_y , line_color )                  
+            #axes.set_aspect('equal')
```
Comment (planes) [inline trunk/SUAVE/Plots/Geometry_Plots/plot_airfoil.py:65]: remove commented out code

File: trunk/SUAVE/Plots/Geometry_Plots/plot_airfoil.py [hunk 9]
```diff
@@ -33,20 +33,38 @@ def plot_airfoil(airfoil_names,  line_color = 'k-', save_figure = False, save_fi
...
-        axes.set_title(airfoil_names[i])
-        axes.plot(airfoil_x, airfoil_y , line_color )                  
-        #axes.set_aspect('equal')
-        axes.axis('equal')
+        for i in range(len(airfoil_names)):
+            # separate x and y coordinates 
+            airfoil_x  = airfoil_data.x_coordinates[i] 
+            airfoil_y  = airfoil_data.y_coordinates[i]    
+            axes.plot(airfoil_x, airfoil_y , label=airfoil_names[i] )                  
+            #axes.axis('equal')
```
Comment (planes) [inline trunk/SUAVE/Plots/Geometry_Plots/plot_airfoil.py:47]: remove commented out code

**Review by planes — COMMENTED** [review]
Let me know if you have any questions.

File: trunk/SUAVE/Methods/Geometry/Two_Dimensional/Cross_Section/Airfoil/generate_airfoil_transitions.py (hunk 2 above)
Comment (rachealerhard) [reply trunk/SUAVE/Methods/Geometry/Two_Dimensional/Cross_Section/Airfoil/generate_airfoil_transitions.py:17]: changed to: generate_interpolated_airfoils

File: trunk/SUAVE/Methods/Geometry/Two_Dimensional/Cross_Section/Airfoil/generate_airfoil_transitions.py (hunk 5 above)
Comment (rachealerhard) [reply trunk/SUAVE/Methods/Geometry/Two_Dimensional/Cross_Section/Airfoil/generate_airfoil_transitions.py:44]: Removed. I originally had an additional output to identify the locations of the newly generated airfoils for a fixed transition region.

File: trunk/SUAVE/Methods/Geometry/Two_Dimensional/Cross_Section/Airfoil/generate_airfoil_transitions.py (hunk 6 above)
Comment (rachealerhard) [reply trunk/SUAVE/Methods/Geometry/Two_Dimensional/Cross_Section/Airfoil/generate_airfoil_transitions.py]: done

File: regression/scripts/airfoil_import/generate_airfoil_transitions.py (hunk 1 above)
Comment (rachealerhard) [reply regression/scripts/airfoil_import/generate_airfoil_transitions.py]: fixed

File: regression/scripts/airfoil_import/generate_airfoil_transitions.py [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+def main():
+    
+    airfoils_path = os.path.join(os.path.dirname(__file__), "Polars/")
+    a_labels  = ["Clark_y", "E63"]
+    space     = 0.2 # distance for transition to occur
+    nairfoils = 4   # number of total airfoils
+    
+    a1 = airfoils_path + a_labels[0]+".txt"
+    a2 = airfoils_path + a_labels[1]+ ".txt"
+    new_files = generate_airfoil_transition(a1, a2, space, nairfoils, save_file=True, save_filename="Polars/Transition")
```
[the remark this answers is not in the document]
Comment (rachealerhard) [reply regression/scripts/airfoil_import/generate_airfoil_transitions.py]: fixed

File: trunk/SUAVE/Methods/Geometry/Two_Dimensional/Cross_Section/Airfoil/generate_airfoil_transitions.py (hunk 3 above)
Comment (rachealerhard) [reply trunk/SUAVE/Methods/Geometry/Two_Dimensional/Cross_Section/Airfoil/generate_airfoil_transitions.py]: This is for the format flag line in the import_airfoil_geometry.py script. I put that in as an arbitrary number greater than 1 to signify the new files are written using the Lednicer format

File: trunk/SUAVE/Plots/Geometry_Plots/plot_airfoil.py (hunk 7 above)
Comment (rachealerhard) [reply trunk/SUAVE/Plots/Geometry_Plots/plot_airfoil.py:37]: hmm.. on my end it gives the one plot of overlayed airfoil geometries and then the 4 individual airfoil plots (5 plots total). Let me know if this is still happening for you with the new changes

**Review by planes — APPROVED** [review]
Nice work!

I didn't get any extra blank plots that time.

**Pull request merged**